Name: Madhusudan Hasbe\
PRN: 22070126061\
TY AIML A3

# ASSIGNMENT 3
---------


# Feedforward Neural Network (from Scratch) for Iris Dataset

## 1. Introduction

In this section, we implement a feedforward neural network from scratch using NumPy for the Iris dataset. The Iris dataset consists of 150 samples of iris flowers, with features like sepal length, sepal width, petal length, and petal width, and it has 3 classes (setosa, versicolor, virginica).

We will use the following neural network structure:
- Input layer: 4 features
- Hidden layer: 5 units
- Output layer: 3 units (one for each class)

### Dataset Preprocessing
We'll normalize the features and convert the labels into one-hot encoding.

### Neural Network Implementation (FeedForwardNN)
Below is the implementation of the `FeedForwardNN` class from scratch:


In [1]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Load the iris dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)


In [2]:
# One hot encode the labels
encoder = OneHotEncoder(sparse=False)
y_onehot = encoder.fit_transform(y)

# Standardize the dataset
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_onehot, test_size=0.2, random_state=42)

c:\Python311\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [3]:
# Define the FeedForwardNN class (from scratch)
class FeedForwardNN:
    def __init__(self, input_size, hidden_size, output_size):
        self.weights1 = np.random.randn(input_size, hidden_size)
        self.bias1 = np.zeros((1, hidden_size))
        self.weights2 = np.random.randn(hidden_size, output_size)
        self.bias2 = np.zeros((1, output_size))

    def forward(self, x):
        self.hidden_layer_activation = np.dot(x, self.weights1) + self.bias1
        self.hidden_layer_output = self.sigmoid(self.hidden_layer_activation)
        self.output_layer_activation = np.dot(self.hidden_layer_output, self.weights2) + self.bias2
        predictions = self.sigmoid(self.output_layer_activation)
        return predictions
    
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))
    
    def backward(self, x, y, output, lr):
        output_error = output - y
        d_weights2 = np.dot(self.hidden_layer_output.T, output_error)
        
        hidden_layer_error = np.dot(output_error, self.weights2.T) * self.sigmoid_derivative(self.hidden_layer_activation)
        d_weights1 = np.dot(x.T, hidden_layer_error)
        
        # Update weights and biases
        self.weights1 -= lr * d_weights1
        self.bias1 -= lr * np.sum(hidden_layer_error, axis=0, keepdims=True)
        self.weights2 -= lr * d_weights2
        self.bias2 -= lr * np.sum(output_error, axis=0, keepdims=True)
    
    def sigmoid_derivative(self, x):
        return self.sigmoid(x) * (1 - self.sigmoid(x))

In [4]:
# Train the model
nn = FeedForwardNN(input_size=4, hidden_size=5, output_size=3)

# Training loop
epochs = 500
lr = 0.01

for epoch in range(epochs):
    output = nn.forward(X_train)
    nn.backward(X_train, y_train, output, lr)

In [5]:
# Predictions
predictions = nn.forward(X_test)
predicted_classes = np.argmax(predictions, axis=1)
true_classes = np.argmax(y_test, axis=1)

# Calculate accuracy
accuracy = np.mean(predicted_classes == true_classes)
print(f"Test Accuracy for Iris Dataset (from scratch): {accuracy * 100:.2f}%")

Test Accuracy for Iris Dataset (from scratch): 100.00%


We got 100% Accuracy on our Feed Forward network built from scratch on Iris Dataset!

----------


# Feedforward Neural Network (with TensorFlow and Keras) for Titanic Dataset

## 1. Introduction

In this section, we will implement a feedforward neural network using TensorFlow and Keras for the Titanic dataset. The Titanic dataset contains information about the passengers, including their survival status.

### Dataset Preprocessing
We will preprocess the data by handling missing values, encoding categorical features, and normalizing the data.

### Neural Network Implementation (with TensorFlow/Keras)
We'll build a simple feedforward neural network using Keras with the following structure:
- Input layer: corresponds to the number of features after preprocessing
- Hidden layer: 16 units with ReLU activation
- Output layer: 1 unit with sigmoid activation for binary classification (survived or not).


In [6]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [7]:
# Load the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
data = pd.read_csv(url)

# Preprocess the data
data = data[['Pclass', 'Sex', 'Age', 'Fare', 'Survived']]
data['Age'].fillna(data['Age'].mean(), inplace=True)

# Convert categorical features
data['Sex'] = LabelEncoder().fit_transform(data['Sex'])

In [8]:
# Separate features and labels
X = data.drop('Survived', axis=1)
y = data['Survived']

# Standardize the data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

In [9]:
# Build a feedforward neural network using Keras
model = Sequential([
    Dense(16, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(1, activation='sigmoid')
])

In [10]:
# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=50, batch_size=32)

Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 0.7242 - accuracy: 0.5646
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 0.6815 - accuracy: 0.6433
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 0.6462 - accuracy: 0.6685
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 0.6166 - accuracy: 0.6896
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 0.5904 - accuracy: 0.7205
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 0.5682 - accuracy: 0.7416
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 0.5482 - accuracy: 0.7542
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 0.5295 - accuracy: 0.7612
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 0.5136 - accuracy: 0.7711
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 0.5005 - accuracy: 0.7781
Epoch 11/50
23/23 [

In [11]:
# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, y_test)
print(f'Test Accuracy for Titanic Dataset (with TensorFlow/Keras): {test_acc * 100:.2f}%')

6/6 [==============================] - 0s 1ms/step - loss: 0.4426 - accuracy: 0.7765
Test Accuracy for Titanic Dataset (with TensorFlow/Keras): 77.65%


We got 77.65% Accuracy for Titanic Datset using TensorFlow/Keras.